File: discovery/push.go
package discovery

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"net/http"
	"strings"
	"time"

	"github.com/xraph/farp"
)

// PushDiscovery implements ServiceDiscovery by pushing manifests directly
// to a gateway via HTTP. No external service registry needed.
//
// This is the service-side component of push-based discovery.
// The gateway-side component is PushHandler.
type PushDiscovery struct {
	gatewayURL string // base URL, e.g., "http://gateway:9090/_farp/v1"
	client     *http.Client
}

// NewPushDiscovery creates a new push-based discovery client.
// gatewayURL is the base URL of the gateway's FARP push endpoint.
func NewPushDiscovery(gatewayURL string, client *http.Client) *PushDiscovery {
	if client == nil {
		client = &http.Client{Timeout: 10 * time.Second}
	}

	return &PushDiscovery{
		gatewayURL: strings.TrimSuffix(gatewayURL, "/"),
		client:     client,
	}
}

// PushRegistration is the body sent when a service pushes its manifest to the gateway.
type PushRegistration struct {
	Instance ServiceInstance      `json:"instance"`
	Manifest *farp.SchemaManifest `json:"manifest"`
}

// PushHeartbeat is the body sent for heartbeat updates.
// RoutesChecksum is optional (§17.4.1 reconciliation).
type PushHeartbeat struct {
	Status         farp.InstanceStatus `json:"status"`
	RoutesChecksum string              `json:"routes_checksum,omitempty"`
}

// PushRegistrationResponse is the gateway's response to a registration.
type PushRegistrationResponse struct {
	Status         string `json:"status"`
	RoutesChecksum string `json:"routes_checksum"`
	SchemasApplied int    `json:"schemas_applied"`
}

// PushHeartbeatResponse is the gateway's response to a heartbeat.
type PushHeartbeatResponse struct {
	Status         string `json:"status"`
	RoutesChecksum string `json:"routes_checksum"`
	SchemasApplied int    `json:"schemas_applied"`
}

// Register pushes the service instance to the gateway.
func (p *PushDiscovery) Register(ctx context.Context, instance ServiceInstance) error {
	_, err := p.doRegister(ctx, PushRegistration{Instance: instance})
	return err
}

// RegisterWithManifest pushes both the instance and manifest to the gateway.
// Use this when the gateway needs the manifest directly (e.g. after a
// heartbeat reconciliation mismatch — see §17.4.1).
func (p *PushDiscovery) RegisterWithManifest(ctx context.Context, instance ServiceInstance, manifest *farp.SchemaManifest) (*PushRegistrationResponse, error) {
	return p.doRegister(ctx, PushRegistration{Instance: instance, Manifest: manifest})
}

func (p *PushDiscovery) doRegister(ctx context.Context, payload PushRegistration) (*PushRegistrationResponse, error) {
	body, err := json.Marshal(payload)
	if err != nil {
		return nil, fmt.Errorf("%w: %w", farp.ErrRegistrationFailed, err)
	}

	req, err := http.NewRequestWithContext(ctx, http.MethodPost, p.gatewayURL+"/register", bytes.NewReader(body))
	if err != nil {
		return nil, fmt.Errorf("%w: %w", farp.ErrRegistrationFailed, err)
	}

	req.Header.Set("Content-Type", "application/json")

	resp, err := p.client.Do(req)
	if err != nil {
		return nil, fmt.Errorf("%w: %w", farp.ErrRegistrationFailed, err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK && resp.StatusCode != http.StatusCreated {
		return nil, fmt.Errorf("%w: HTTP %d", farp.ErrRegistrationFailed, resp.StatusCode)
	}

	var ack PushRegistrationResponse
	if err := json.NewDecoder(resp.Body).Decode(&ack); err != nil {
		// Gateway may not support ack yet — treat as success with empty state
		return &PushRegistrationResponse{Status: "registered"}, nil
	}

	return &ack, nil
}

// Deregister removes the service from the gateway.
func (p *PushDiscovery) Deregister(ctx context.Context, instanceID string) error {
	req, err := http.NewRequestWithContext(ctx, http.MethodDelete, p.gatewayURL+"/deregister/"+instanceID, nil)
	if err != nil {
		return fmt.Errorf("%w: %w", farp.ErrDeregistrationFailed, err)
	}

	resp, err := p.client.Do(req)
	if err != nil {
		return fmt.Errorf("%w: %w", farp.ErrDeregistrationFailed, err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK && resp.StatusCode != http.StatusNoContent {
		return fmt.Errorf("%w: HTTP %d", farp.ErrDeregistrationFailed, resp.StatusCode)
	}

	return nil
}

// ReportHealth sends a heartbeat to the gateway (without checksum).
// Satisfies the ServiceDiscovery interface.
func (p *PushDiscovery) ReportHealth(ctx context.Context, instanceID string, status farp.InstanceStatus) error {
	_, err := p.ReportHealthWithChecksum(ctx, instanceID, status, "")
	return err
}

// ReportHealthWithChecksum sends a heartbeat with the service's expected
// routes_checksum for reconciliation (FARP spec §17.4.1).
// The response includes the gateway's applied checksum and schema count.
func (p *PushDiscovery) ReportHealthWithChecksum(ctx context.Context, instanceID string, status farp.InstanceStatus, routesChecksum string) (*PushHeartbeatResponse, error) {
	payload := PushHeartbeat{
		Status:         status,
		RoutesChecksum: routesChecksum,
	}

	body, err := json.Marshal(payload)
	if err != nil {
		return nil, fmt.Errorf("%w: %w", farp.ErrHealthCheckFailed, err)
	}

	req, err := http.NewRequestWithContext(ctx, http.MethodPut, p.gatewayURL+"/heartbeat/"+instanceID, bytes.NewReader(body))
	if err != nil {
		return nil, fmt.Errorf("%w: %w", farp.ErrHealthCheckFailed, err)
	}

	req.Header.Set("Content-Type", "application/json")

	resp, err := p.client.Do(req)
	if err != nil {
		return nil, fmt.Errorf("%w: %w", farp.ErrHealthCheckFailed, err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("%w: HTTP %d", farp.ErrHealthCheckFailed, resp.StatusCode)
	}

	var ack PushHeartbeatResponse
	if err := json.NewDecoder(resp.Body).Decode(&ack); err != nil {
		// Gateway may not support ack yet
		return &PushHeartbeatResponse{Status: "ok"}, nil
	}

	return &ack, nil
}

// Discover queries the gateway for instances of a service.
func (p *PushDiscovery) Discover(ctx context.Context, serviceName string) ([]ServiceInstance, error) {
	url := p.gatewayURL + "/services"
	if serviceName != "" {
		url += "/" + serviceName
	}

	req, err := http.NewRequestWithContext(ctx, http.MethodGet, url, nil)
	if err != nil {
		return nil, err
	}

	resp, err := p.client.Do(req)
	if err != nil {
		return nil, fmt.Errorf("%w: %w", farp.ErrDiscoveryUnavailable, err)
	}
	defer resp.Body.Close()

	var instances []ServiceInstance
	if err := json.NewDecoder(resp.Body).Decode(&instances); err != nil {
		return nil, err
	}

	return instances, nil
}

// Watch is not natively supported in push mode from the service side.
// Services push to the gateway; they don't watch it.
// This is a no-op that blocks until context is cancelled.
func (p *PushDiscovery) Watch(ctx context.Context, _ string, _ DiscoveryEventHandler) error {
	<-ctx.Done()
	return nil
}

// Close is a no-op for push discovery.
func (p *PushDiscovery) Close() error {
	return nil
}

// Health checks if the gateway is reachable.
func (p *PushDiscovery) Health(ctx context.Context) error {
	req, err := http.NewRequestWithContext(ctx, http.MethodGet, p.gatewayURL+"/services", nil)
	if err != nil {
		return err
	}

	resp, err := p.client.Do(req)
	if err != nil {
		return fmt.Errorf("%w: %w", farp.ErrDiscoveryUnavailable, err)
	}
	defer resp.Body.Close()

	if resp.StatusCode >= 500 {
		return fmt.Errorf("%w: HTTP %d", farp.ErrDiscoveryUnavailable, resp.StatusCode)
	}

	return nil
}

File: farp-ts/src/discovery/push.ts
/**
 * PushDiscovery implements ServiceDiscovery by pushing manifests directly
 * to a gateway via HTTP. No external service registry needed.
 *
 * This is the service-side component of push-based discovery.
 * The gateway-side component is PushHandler.
 *
 * Ported from Go: discovery/push.go
 */

import type { InstanceStatus, SchemaManifest } from '../types';
import {
  RegistrationFailedError,
  DeregistrationFailedError,
  HealthCheckFailedError,
  DiscoveryUnavailableError,
} from '../errors';
import type { ServiceDiscovery } from './interface';
import type {
  ServiceInstance,
  DiscoveryEventHandler,
  PushRegistration,
  PushHeartbeat,
  PushRegistrationResponse,
  PushHeartbeatResponse,
} from './types';

/**
 * Type for the fetch function, allowing injection for testing or custom runtimes.
 */
type FetchFn = typeof globalThis.fetch;

/**
 * PushDiscovery implements ServiceDiscovery by pushing manifests directly
 * to a gateway via HTTP. No external service registry needed.
 */
export class PushDiscovery implements ServiceDiscovery {
  private readonly gatewayURL: string;
  private readonly fetchFn: FetchFn;
  private abortController: AbortController | null = null;

  /**
   * Creates a new push-based discovery client.
   * @param gatewayURL - Base URL of the gateway's FARP push endpoint (e.g. "http://gateway:9090/_farp/v1")
   * @param fetchFn - Optional custom fetch function (defaults to globalThis.fetch)
   */
  constructor(gatewayURL: string, fetchFn?: FetchFn) {
    this.gatewayURL = gatewayURL.replace(/\/+$/, '');
    this.fetchFn = fetchFn ?? globalThis.fetch.bind(globalThis);
  }

  /**
   * Register pushes the service instance to the gateway.
   */
  async register(instance: ServiceInstance): Promise<void> {
    await this.doRegister({ instance });
  }

  /**
   * RegisterWithManifest pushes both the instance and manifest to the gateway.
   * Use this when the gateway needs the manifest directly (e.g. after a
   * heartbeat reconciliation mismatch -- see section 17.4.1).
   */
  async registerWithManifest(
    instance: ServiceInstance,
    manifest: SchemaManifest,
  ): Promise<PushRegistrationResponse> {
    return this.doRegister({ instance, manifest });
  }

  private async doRegister(
    payload: PushRegistration,
  ): Promise<PushRegistrationResponse> {
    let body: string;
    try {
      body = JSON.stringify(payload);
    } catch (err) {
      throw new RegistrationFailedError(
        `failed to serialize registration: ${err}`,
      );
    }

    let resp: Response;
    try {
      resp = await this.fetchFn(`${this.gatewayURL}/register`, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body,
      });
    } catch (err) {
      throw new RegistrationFailedError(
        `fetch failed: ${err}`,
      );
    }

    if (resp.status !== 200 && resp.status !== 201) {
      throw new RegistrationFailedError(
        `HTTP ${resp.status}`,
      );
    }

    try {
      const ack: PushRegistrationResponse = await resp.json();
      return ack;
    } catch {
      // Gateway may not support ack yet -- treat as success with empty state
      return { status: 'registered', routes_checksum: '', schemas_applied: 0 };
    }
  }

  /**
   * Deregister removes the service from the gateway.
   */
  async deregister(instanceId: string): Promise<void> {
    let resp: Response;
    try {
      resp = await this.fetchFn(
        `${this.gatewayURL}/deregister/${instanceId}`,
        { method: 'DELETE' },
      );
    } catch (err) {
      throw new DeregistrationFailedError(
        `fetch failed: ${err}`,
      );
    }

    if (resp.status !== 200 && resp.status !== 204) {
      throw new DeregistrationFailedError(
        `HTTP ${resp.status}`,
      );
    }
  }

  /**
   * ReportHealth sends a heartbeat to the gateway (without checksum).
   * Satisfies the ServiceDiscovery interface.
   */
  async reportHealth(
    instanceId: string,
    status: InstanceStatus,
  ): Promise<void> {
    await this.reportHealthWithChecksum(instanceId, status, '');
  }

  /**
   * ReportHealthWithChecksum sends a heartbeat with the service's expected
   * routes_checksum for reconciliation (FARP spec section 17.4.1).
   * The response includes the gateway's applied checksum and schema count.
   */
  async reportHealthWithChecksum(
    instanceId: string,
    status: InstanceStatus,
    routesChecksum: string,
  ): Promise<PushHeartbeatResponse> {
    const payload: PushHeartbeat = {
      status,
      routes_checksum: routesChecksum || undefined,
    };

    let body: string;
    try {
      body = JSON.stringify(payload);
    } catch (err) {
      throw new HealthCheckFailedError(
        `failed to serialize heartbeat: ${err}`,
      );
    }

    let resp: Response;
    try {
      resp = await this.fetchFn(
        `${this.gatewayURL}/heartbeat/${instanceId}`,
        {
          method: 'PUT',
          headers: { 'Content-Type': 'application/json' },
          body,
        },
      );
    } catch (err) {
      throw new HealthCheckFailedError(
        `fetch failed: ${err}`,
      );
    }

    if (resp.status !== 200) {
      throw new HealthCheckFailedError(
        `HTTP ${resp.status}`,
      );
    }

    try {
      const ack: PushHeartbeatResponse = await resp.json();
      return ack;
    } catch {
      // Gateway may not support ack yet
      return { status: 'ok', routes_checksum: '', schemas_applied: 0 };
    }
  }

  /**
   * Discover queries the gateway for instances of a service.
   */
  async discover(serviceName?: string): Promise<ServiceInstance[]> {
    let url = `${this.gatewayURL}/services`;
    if (serviceName) {
      url += `/${serviceName}`;
    }

    let resp: Response;
    try {
      resp = await this.fetchFn(url, { method: 'GET' });
    } catch (err) {
      throw new DiscoveryUnavailableError(
        `fetch failed: ${err}`,
      );
    }

    const instances: ServiceInstance[] = await resp.json();
    return instances;
  }

  /**
   * Watch is not natively supported in push mode from the service side.
   * Services push to the gateway; they don't watch it.
   * This blocks until the AbortController is aborted (via close()).
   */
  async watch(
    _serviceName: string | undefined,
    _handler: DiscoveryEventHandler,
  ): Promise<void> {
    this.abortController = new AbortController();
    return new Promise<void>((resolve) => {
      this.abortController!.signal.addEventListener('abort', () => {
        resolve();
      });
    });
  }

  /**
   * Close is a no-op for push discovery (aborts any pending watch).
   */
  close(): void {
    if (this.abortController) {
      this.abortController.abort();
      this.abortController = null;
    }
  }

  /**
   * Health checks if the gateway is reachable.
   */
  async health(): Promise<void> {
    let resp: Response;
    try {
      resp = await this.fetchFn(`${this.gatewayURL}/services`, {
        method: 'GET',
      });
    } catch (err) {
      throw new DiscoveryUnavailableError(
        `fetch failed: ${err}`,
      );
    }

    if (resp.status >= 500) {
      throw new DiscoveryUnavailableError(
        `HTTP ${resp.status}`,
      );
    }
  }
}
